src/main/scala/com/kinja/amqp/AmqpConsumer.scala

```diff
@@ -113,47 +130,82 @@ class Listener[A: Reads](
 	private val toleranceNanos = 10000000L // 10 milliseconds
 
 	/**
-	 * Default state of the listener which receives messages from RabbitMQ.
+	 * receive method of the Listener which receives messages from RabbitMQ, default state is idle.
 	 */
-	def receive = {
-		case Delivery(consumerTag, envelope, properties, body) =>
+	def receive: Receive = idle
+
+	/**
+	 * Idle state of the listener which receives messages from RabbitMQ.
+		*  Listener waiting for new Delivery
+	 */
+	def idle: Receive = {
+		case Delivery(_, envelope, _, body) =>
 			val nextTickNanos = System.nanoTime + spacing.toNanos
 			val s = new String(body, "UTF-8")
 
+			val originalSender = sender
 			implicitly[Reads[A]].reads(s) match {
 				case Right(message) =>
-					try {
-						Await.result(processor(message), timeout)
-						val ack = Ack(envelope.getDeliveryTag)
-
-						// sleep until we are allowd to receive a new message
-						val nowNanos = System.nanoTime
-						if (nowNanos < nextTickNanos - toleranceNanos) {
-							implicit val ec: ExecutionContext = context.dispatcher
-							ignore(context.system.scheduler.scheduleOnce((nextTickNanos - nowNanos).nanos, self, WakeUp))
-							context.become(asleep(sender, ack))
-						} else {
-							sender ! ack
-						}
-					} catch {
-						case NonFatal(t) =>
-							logger.warn(s"""[RabbitMQ] Exception while processing message "$s" : $t""")
-							sender ! Reject(envelope.getDeliveryTag, requeue = true)
-					}
+					context.become(processing, true)
```

Comment:
Probably would be better to have `originalSender` and `envelope` as parameters to `processing`. Then you don't need them in the processing result message, and it becomes much easier.